# Assignment 3 UDS

*   Arden Silva
*   Neetu Singh
*   Pamela Nunez
*   Shriraksha Srilakshmi Nataraj
*Sri Amirdha Sudha

# TASK A

Scrape Instagram.py to fetch ~500 posts from the brand’s Instagram page (We chose "Nike").
This script fetches:

(i) image URLs,

(ii) post caption (the text description of a post)

(iii) number of likes.

In [ ]:
# pip install
pip install instaloader


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.1/65.1 kB 1.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for instaloader: filename=instaloader-4.13.1-py3-none-any.whl size=67215 sha256=4971ca586dbb5752de2c936e237e5259cde1c7cfc100e6946f8a2f4d8163349a
  Stored in directory: /root/.cache/pip/wheels/be/53/a0/8d9867ecfa8dbb1f7a6f17be76c2c27836218843653a92e80c
Successfully built instaloader


In [ ]:
#imports
import instaloader
import pandas as pd

# Initialize Instaloader
L = instaloader.Instaloader()

# Log in to Instagram (if needed)
L.login('alpha_wolf', 'dkbose123')

# load nike profile
profile = instaloader.Profile.from_username(L.context, 'nike')

# Create lists to store scraped data
image_urls = []
post_captions = []
likes = []

# Fetch the first 500 posts
for post in profile.get_posts():
    if len(image_urls) >= 500:  # Limit to 500 posts
        break
    image_urls.append(post.url)
    post_captions.append(post.caption)
    likes.append(post.likes)

# Create a pandas DataFrame
data = pd.DataFrame({
    'image_url': image_urls,
    'post_caption': post_captions,
    'likes': likes
})

# Save the data to a CSV file for further use
data.to_csv('nike_instagram_posts.csv', index=False)

# Display the first few rows of the dataset
print(data.head())


                                           image_url  \
0  https://scontent-iad3-2.cdninstagram.com/v/t51...   
1  https://scontent-iad3-2.cdninstagram.com/v/t51...   
2  https://scontent-iad3-2.cdninstagram.com/v/t51...   
3  https://scontent-iad3-2.cdninstagram.com/v/t51...   
4  https://scontent-iad3-2.cdninstagram.com/v/t51...   

                                        post_caption   likes  
0  “You never get used to pressure. You just lear...  133526  
1  Let’s get something straight. Winning is winni...   49362  
2  Winning Isn’t For Everyone. It’s for those who...  249478  
3       A monumental statement with a ring to match.  276865  
4  Pressure tested. Pressure ready. \n\nThe LeBro...  187958  


# TASK B

In [ ]:
#imports
import pandas as pd

#load the csv
df = pd.read_csv('nike_instagram_posts.csv')

In [ ]:
#must install google cloud vision
!pip install google-cloud-vision


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 467.5/467.5 kB 6.0 MB/s eta 0:00:00


In [ ]:
# imports
import os
from google.cloud import vision
import requests

# get environment to secret
os.environ["GOOGLE_APPLICATION_CREDENTIALS"] =  '/content/keen-virtue-419219-742db1c7d611.json'

# Initialize the Vision client
client = vision.ImageAnnotatorClient()

#df method to get labels
def extract_labels(url): #takes img_url column
    image = vision.Image(content=requests.get(url).content)  # Fetch image content
    response = client.label_detection(image=image) #get labels
    labels = [label.description for label in response.label_annotations] #put in list format
    return labels #return labels

df['labels'] = df['image_url'].apply(extract_labels) #apply function in new column called labels

In [ ]:
df

image_url  \
Unnamed: 0                                                      
0           https://scontent-iad3-2.cdninstagram.com/v/t51...   
1           https://scontent-iad3-2.cdninstagram.com/v/t51...   
2           https://scontent-iad3-2.cdninstagram.com/v/t51...   
3           https://scontent-iad3-2.cdninstagram.com/v/t51...   
4           https://scontent-iad3-2.cdninstagram.com/v/t51...   
...                                                       ...   
495         https://scontent-iad3-2.cdninstagram.com/v/t51...   
496         https://scontent-iad3-1.cdninstagram.com/v/t51...   
497         https://scontent-iad3-1.cdninstagram.com/v/t51...   
498         https://scontent-iad3-1.cdninstagram.com/v/t51...   
499         https://scontent-iad3-1.cdninstagram.com/v/t51...   

                                                 post_caption   likes  \
Unnamed: 0                                                              
0           “You never get used to pressure. You just lear...  133526   
1           Let’s get something straight. Winning is winni...   49362   
2           Winning Isn’t For Everyone. It’s for those who...  249478   
3                A monumental statement with a ring to match.  276865   
4           Pressure tested. Pressure ready. \n\nThe LeBro...  187958   
...                                                       ...     ...   
495         CODE 05296: “CULTURE”\n\nNever forget where yo...   92561   
496         You can’t be what you can’t see. And for footb...   86785   
497         CODE 00099: “SIGNATURE”\nDo what is unexpected...   86022   
498         The Drake AF1 ‘Love You Forever’. 💕\n\nA refre...  237498   
499         @ibtihajmuhammad was the first Muslim American...   76237   

                                                       labels  
Unnamed: 0                                                     
0           ['Forehead', 'Nose', 'Chin', 'Hairstyle', 'Eye...  
1           ['Forehead', 'Nose', 'Hair', 'Cheek', 'Skin', ...  
2           ['Forehead', 'Nose', 'Cheek', 'Lip', 'Eyelash'...  
3           ['Sculpture', 'Finial', 'Gesture', 'Statue', '...  
4           ['Footwear', 'Walking shoe', 'Sportswear', 'Sn...  
...                                                       ...  
495         ['Jaw', 'Sleeve', 'Vision care', 'Facial hair'...  
496         ['Forehead', 'Sleeve', 'Tableware', 'Drinkware...  
497         ['Hat', 'Sleeve', 'Sun hat', 'Font', 'Helmet',...  
498         ['Wheel', 'Tire', 'Bicycle', 'Bicycles--Equipm...  
499         ['Smile', 'Tableware', 'Textile', 'Leisure', '...  

[500 rows x 4 columns]

In [ ]:
df['labels'][:5]

0    ['Forehead', 'Nose', 'Chin', 'Hairstyle', 'Eye...
1    ['Forehead', 'Nose', 'Hair', 'Cheek', 'Skin', ...
2    ['Forehead', 'Nose', 'Cheek', 'Lip', 'Eyelash'...
3    ['Sculpture', 'Finial', 'Gesture', 'Statue', '...
4    ['Footwear', 'Walking shoe', 'Sportswear', 'Sn...
Name: labels, dtype: object

# TASK C

Task C: Create a column called binary (lowercase only) where value =1 (stands for high engagement) or 0 (stands for low engagement) based on whether the number of likes is above or below the median value.

In [ ]:
# creates a binary column
# if likes are greater than the median, return as 1, and 0 otherwise
df['binary'] = (df['likes'] > df['likes'].median()).astype(int) #this converts it into 1 and 0s instead of true and false
df[['likes','binary']][:5] #check to make sure its working

likes  binary
Unnamed: 0                
0           133526       1
1            49362       0
2           249478       1
3           276865       1
4           187958       1

In [ ]:
#save as a df

df.to_csv('nike_instagram_posts_with_labels.csv')

# TASK D
Run a logistic regression with binary as the dependent variable, and the image_labels as independent variables. Before running the regression, replace the column label image_labels with text, since the script expects text to be the name for the column containing text. What is the accuracy (show the confusion matrix) of this prediction model? The idea is to be able to predict the engagement level for an image.
Accuracy = 1 - # prediction errors / total # cases
What accuracy do you get by using the post_caption words as the independent variables instead of image_labels? As in the first regression, change post_caption to text before running the logistic regression. Finally, what accuracy do you get by combining (concatenating) the image_labels and post_caption and using them together as independent variables? What can you conclude from your analysis?

**Data Preparation**

In [ ]:
# Imports
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import confusion_matrix, accuracy_score
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem.snowball import SnowballStemmer

# Download NLTK resources for text preprocessing
nltk.download('stopwords')
nltk.download('punkt')

# Initialize a stemmer and load English stop words
stemmer = SnowballStemmer("english")
stop_words = set(stopwords.words('english'))

# Function to process text by tokenizing, removing stop words, and stemming
def preprocess_text(text):
    tokens = word_tokenize(text.lower())  # Convert text to lowercase and tokenize it
    # Remove non-alphabetical words and stop words, then apply stemming
    filtered_tokens = [stemmer.stem(word) for word in tokens if word.isalpha() and word not in stop_words]
    return " ".join(filtered_tokens)  # Join the processed words back into a string

# Load the dataset containing Instagram post data
df = pd.read_csv('nike_instagram_posts_with_labels.csv')

# Create a binary column based on the median value of 'likes'
median_likes = df['likes'].median()  # Calculate the median likes
# Assign 1 if the number of likes is greater than or equal to the median, otherwise 0
df['binary'] = (df['likes'] >= median_likes).astype(int)

# Process and clean 'labels' and 'post_caption' columns
df['labels_text'] = df['labels'].apply(lambda x: ' '.join(eval(x)))
df['post_caption'] = df['post_caption'].fillna('')

# Apply the text preprocessing function to the 'labels_text' and 'post_caption' columns
df['post_caption'] = df['post_caption'].apply(preprocess_text)
df['labels_text'] = df['labels_text'].apply(preprocess_text)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


**Modeling**

In [ ]:
# Vectorize the text columns using TF-IDF (Term Frequency-Inverse Document Frequency)

# for 'labels_text' column
vectorizer_labels = TfidfVectorizer(stop_words='english')
X_labels = vectorizer_labels.fit_transform(df['labels_text'])

# for 'post_caption' column
vectorizer_caption = TfidfVectorizer(stop_words='english')
X_caption = vectorizer_caption.fit_transform(df['post_caption'])

# Combine 'labels_text' and 'post_caption' columns into a single feature for combined analysis
combined_text = df['labels_text'] + ' ' + df['post_caption']
X_combined = vectorizer_caption.fit_transform(combined_text)

# Split the data into training and testing sets for each feature set (80% training, 20% testing)
X_train_labels, X_test_labels, y_train, y_test = train_test_split(X_labels, df['binary'], test_size=0.2, random_state=42)
X_train_caption, X_test_caption, _, _ = train_test_split(X_caption, df['binary'], test_size=0.2, random_state=42)
X_train_combined, X_test_combined, _, _ = train_test_split(X_combined, df['binary'], test_size=0.2, random_state=42)

# Initialize a Logistic Regression model
logreg = LogisticRegression(max_iter=1000)

# Train and evaluate the model using 'labels' only
logreg.fit(X_train_labels, y_train)
y_pred_labels = logreg.predict(X_test_labels)

# Train and evaluate the model using 'post_caption' only
logreg.fit(X_train_caption, y_train)
y_pred_caption = logreg.predict(X_test_caption)

# Train and evaluate the model using combined 'labels' and 'post_caption'
logreg.fit(X_train_combined, y_train)
y_pred_combined = logreg.predict(X_test_combined)

In [ ]:
y_pred_combined

array([1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0,
       0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0])

**Evaluation**

In [ ]:
# Function to evaluate the model using confusion matrix and accuracy score
def evaluate_model(y_test, y_pred, label):
    conf_matrix = confusion_matrix(y_test, y_pred)  # Calculate confusion matrix
    accuracy = accuracy_score(y_test, y_pred)  # Calculate accuracy score
    print(f"Results using {label}:")
    print("Confusion Matrix:")
    print(conf_matrix)
    print("Accuracy:", accuracy)

# Evaluate each case (using 'labels' only, 'post_caption' only, and both combined)
evaluate_model(y_test, y_pred_labels, "'labels' only")
evaluate_model(y_test, y_pred_caption, "'post_caption' only")
evaluate_model(y_test, y_pred_combined, "combined 'labels' and 'post_caption'")

Results using 'labels' only:
Confusion Matrix:
[[32 17]
 [23 28]]
Accuracy: 0.6
Results using 'post_caption' only:
Confusion Matrix:
[[33 16]
 [23 28]]
Accuracy: 0.61
Results using combined 'labels' and 'post_caption':
Confusion Matrix:
[[32 17]
 [21 30]]
Accuracy: 0.62


**Results:**

**Using Labels Only:**

Confusion Matrix:
<br>[[32, 17],
<br>[23, 28]]
<br>The accuracy is 60%

**Using Post Captions Only:**

Confusion Matrix:
<br>[[33, 16],
<br>[23, 28]]
<br>The accuracy is 61%

**Using Combined Labels and Captions:**

Confusion Matrix:
<br>[[32, 17],
<br>[21, 30]]
<br>The accuracy is 62%

**Observations:**
Captions seem to be more important for predicting engagement than labels.
Combining both labels and captions gives a little more accuracy, showing that both visual and text elements play a role in engaging users.
However, the increase in accuracy when combining both is not very large, suggesting that captions carry most of the information needed for predicting engagement.

# TASK E

 Perform topic modeling (LDA) on the original image_labels. Choose an appropriate
number of topics. You may want to start with 4-5 topics, but adjust the number up or down
depending on the word distributions you get. Decide on suitable names for each topic.
Now sort the data from high to low number of likes (don’t use the binary column, use the actual
number of likes), and consider the highest and the lowest quartiles of likes. What are the main
differences in the average topic weights of images across the two quartiles (e.g., greater weight
of some topics in the highest versus lowest quartiles)? Show the main results in a table.

In [ ]:
!pip install lda

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 350.0/350.0 kB 4.8 MB/s eta 0:00:00


In [ ]:
# Import necessary libraries
import os
import csv
import nltk
import lda
import pandas as pd
import numpy as np
from nltk.tokenize import RegexpTokenizer
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import CountVectorizer
# Download required NLTK resources if not already available
nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [ ]:

# Load the data from an Excel file into a DataFrame
df = pd.read_csv('nike_instagram_posts_with_labels.csv').set_index('Unnamed: 0')
# Check for null values in the DataFrame and print the count of empty columns
df.isnull().sum() #it looks like we have a null post_caption

image_url       0
post_caption    1
likes           0
labels          0
binary          0
dtype: int64

In [ ]:
df[df['post_caption'].isna()] #this is the row, lets drop it

image_url post_caption  \
Unnamed: 0                                                                   
129         https://scontent-iad3-2.cdninstagram.com/v/t51...          NaN   

             likes                                             labels  binary  
Unnamed: 0                                                                     
129         649095  ['Forehead', 'Nose', 'Eyebrow', 'Beard', 'Jaw'...       1

In [ ]:
# Drop rows with any null values
df = df.dropna()

#reset index
df = df.reset_index()
df = df.drop(columns='index')
df.index.names = ['index']

In [ ]:
# Initialize tokenizer and lemmatizer
tokenizer = RegexpTokenizer(r'\w+')
lemmatizer = WordNetLemmatizer()

# Set of English stopwords
stop_words = set(stopwords.words('english'))


#Function to tokenize
# Tokenizes text by words
# Lemmatizes words to their base form
def preprocess_txt(text): #takes text column as an input
  #tokenize
  tokens = tokenizer.tokenize(text.lower())
  #take out stop words
  tokens = [word for word in tokens if word not in stop_words]
  #lemmatize
  tokens = [lemmatizer.lemmatize(word) for word in tokens]
  return tokens #return

# Create a CountVectorizer to transform text data into a matrix of token counts
vectorizer = CountVectorizer(analyzer=preprocess_txt)

# Fit and transform the text
X = vectorizer.fit_transform(df['text'])


In [ ]:
# Print the shape of the feature matrix
X.shape

(499, 564)

In [ ]:
input_topics = 3
# Initialize and fit the LDA model
model = lda.LDA(n_iter=1000,n_topics=input_topics)
#fit
model.fit(X)

In [ ]:
#  Extract the topic-word distributio
topic_word_dist = model.topic_word_

# Get the vocabulary from the CountVectorizer
vocab = vectorizer.get_feature_names_out()

# Extract words for each topic
for k in range(input_topics): #for each topic
  topic_0_words = np.argsort(topic_word_dist[k])[::-1][:10]  # Get top 10 words
  topic_0_words_str = [vocab[i] for i in topic_0_words]

  print(f"Topic {k+1} words: {topic_0_words_str}")

Topic 1 words: ['sport', 'player', 'equipment', 'uniform', 'jersey', 'short', 'sportswear', 'gesture', 'ball', 'basketball']
Topic 2 words: ['shoe', 'sleeve', 'human', 'photography', 'leg', 'flash', 'thigh', 'gesture', 'footwear', 'sportswear']
Topic 3 words: ['art', 'happy', 'sky', 'automotive', 'event', 'design', 'photography', 'flash', 'leisure', 'equipment']


We have come to around 3 topics using our dataset.

- Topic 1: athletics/athletes
- Topic 2: shoes
- Topic 3: artistic photography (graphics)


In [ ]:
# Extract document-topic distribution from the model
doc_topic_dist = model.doc_topic_

# Convert the document-topic distribution to a DataFrame
doc_topic_df = pd.DataFrame(doc_topic_dist, columns=['athletics', 'shoes', 'photography']).reset_index()

df = df.join(doc_topic_df, on='index',lsuffix='idx')


In [ ]:
df[:5]

level_0                                          image_url  \
index                                                               
0            0  https://scontent-iad3-2.cdninstagram.com/v/t51...   
1            1  https://scontent-iad3-2.cdninstagram.com/v/t51...   
2            2  https://scontent-iad3-2.cdninstagram.com/v/t51...   
3            3  https://scontent-iad3-2.cdninstagram.com/v/t51...   
4            4  https://scontent-iad3-2.cdninstagram.com/v/t51...   

                                            post_caption   likes  \
index                                                              
0      “You never get used to pressure. You just lear...  133526   
1      Let’s get something straight. Winning is winni...   49362   
2      Winning Isn’t For Everyone. It’s for those who...  249478   
3           A monumental statement with a ring to match.  276865   
4      Pressure tested. Pressure ready. \n\nThe LeBro...  187958   

                                                  labels  binary  \
index                                                              
0      ['Forehead', 'Nose', 'Chin', 'Hairstyle', 'Eye...       1   
1      ['Forehead', 'Nose', 'Hair', 'Cheek', 'Skin', ...       0   
2      ['Forehead', 'Nose', 'Cheek', 'Lip', 'Eyelash'...       1   
3      ['Sculpture', 'Finial', 'Gesture', 'Statue', '...       1   
4      ['Footwear', 'Walking shoe', 'Sportswear', 'Sn...       1   

                                                    text  index  athletics  \
index                                                                        
0      Forehead Nose Chin Hairstyle Eyebrow Facial ex...      0   0.008130   
1      Forehead Nose Hair Cheek Skin Head Lip Chin Ey...      1   0.009709   
2      Forehead Nose Cheek Lip Eyelash Human body Nec...      2   0.008130   
3      Sculpture Finial Gesture Statue Art Gas Artifa...      3   0.009709   
4      Footwear Walking shoe Sportswear Sneakers Outd...      4   0.006135   

          shoes  photography  
index                         
0      0.983740     0.008130  
1      0.980583     0.009709  
2      0.983740     0.008130  
3      0.300971     0.689320  
4      0.987730     0.006135

In [ ]:
#sort from high to low
df = df.sort_values(by='likes',ascending=False)

#get quartiles
q1 = df['likes'].quantile(0.25)
q3 = df['likes'].quantile(0.75)

#get posts
df_high_likes = df[df['likes'] >= q3]
df_low_likes = df[df['likes'] <= q1]


In [ ]:
#get averages
avg_topics_high = df_high_likes[['athletics', 'shoes', 'photography']].mean(axis=0)
avg_topics_low = df_low_likes[['athletics', 'shoes', 'photography']].mean(axis=0)

#make a table to show
comparison_table = pd.DataFrame({'Quartile': ['High Likes', 'Low Likes'], #shoes quartile
                                  'Athletics': [avg_topics_high['athletics'], avg_topics_low['athletics']], #athletics mean
                                  'Shoes': [avg_topics_high['shoes'], avg_topics_low['shoes']], #shoes mean
                                  'Photography': [avg_topics_high['photography'], avg_topics_low['photography']]}) #photography mean

comparison_table

Quartile  Athletics     Shoes  Photography
0  High Likes   0.292292  0.443581     0.264128
1   Low Likes   0.338684  0.383205     0.278110

# Task F

What advice would you give to the brand if it wants to increase engagement on its
Instagram page based on your findings?

From the findings and observation of previous tasks, these are our insights:


1. As per the topics picked by the topic modeling, the shoe topic has the highest engagement (0.44) and (0.38), reflected in the higher number of likes. Nike has to focus more on its footwear and sportswear offerings. The shoe topic mainly includes words like footwear, sportswear and other shoes. Nike can use this strategy to enhance customer engagement on Instagram by endorsing their latest shoe catalog. They can even use this social platform to expand their footprint by providing discounts on frequently bought sportswear. By giving out offers can attract more customers who can afford that price. They can also include their best-selling shoe image on their partnerships pages which can increase their user engagement and eventually reach a wider audience digitally.

2. On the other hand, they should also look at the least number of likes that are generated from the topic photography. This mostly includes images of nature, humans, art, design, and equipment. This is the area Nike can spend less time and efforts on if they are continuing to post same image. If they enhance and try to use more of Nike shoes and sportswear in these images, that could yield them more likes. They can focus more on their offerings and capture the images with the best quality. They should reduce the including of art, design, and types of equipment on their images, as people are not paying attention to that. Customers are more interested in their offerings than in other artistic designs with which they broadcast their new product.

3. Based on the Logistic Regression model results, we see that the accuracy is mostly similar for the posts, considering labels, caption and both separately as independent variable to predict the model. However, looking at the quantitative result, Nike should be focussing on the captions and labels of the image they are broadcasting. With 62% accuracy for both (labels and captions) as independent variables, we suggest that Nike should consider what caption they post with that image.

4. Using Image Analytics and Machine Learning Algorithms, Nike can distinguish between their least and highest number of likes associated with the images. This information can assist them in deciding their marketing with the right strategy.

